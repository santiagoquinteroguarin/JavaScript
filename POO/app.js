// ?class declaration
class Client {

    // ?constructor method
    constructor(name, money) {
        this.name = name;
        this.money = money;
    }

    // ?getters and setters
    setName(name) {
        this.name = name;
    }

    getName() {
        return this.name;
    }

    setMoney(money) {
        this.money = money;
    }

    getMoney() {
        return this.money;
    }

    // ?methods
    printData() {
        return `Cliente: ${this.getName()}, tu money es: ${this.getMoney()}`;
    }

    // ?static properties
    static welcome() {
        return `Welcome To ATM`;
    }
}

// ?Class inheritance
class Company extends Client {

    // ?private properties
    #category;

    constructor(name, money, phone, category) {
        super(name, money);
        this.phone = phone;
        this.#category = category;
    }

    setPhone(phone) {
        this.phone = phone;
    }

    getPhone() {
        return this.phone;
    }   

    setCategory(category) {
        this.#category = category;
    }

    getCategory() {
        return this.#category;
    }

    infoCompany() {
        return `Company: ${this.getName()}, Category: ${this.getCategory()}, and the contact is: ${this.getPhone()}`;
    }

    // ?static properties
    static welcome() {
        return `Welcome To ATM of Companies`;
    } 
}

const santiago = new Client();
const sugarCompany = new Company();


santiago.setName('Santiago');
santiago.setMoney(5000);

sugarCompany.setName('Sugar Company');
sugarCompany.setPhone(123456789);
sugarCompany.setMoney(20000000);
sugarCompany.setCategory('Candies');

console.log(Company.welcome());
console.log(sugarCompany.infoCompany())

console.log(Client.welcome());
console.log(santiago.printData());
