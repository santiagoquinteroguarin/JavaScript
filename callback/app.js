const countries = [];

function newCountry(country, callback) {
    countries.push(country);
    console.log(`attche: ${country}`);
    callback();
}

function printCountries() {
    console.log(countries);
}

function startCallbackHell() {
    setTimeout(() => {
        newCountry('Colombia', printCountries);

        setTimeout(() => {
            newCountry('Mexico', printCountries);

            setTimeout(() => {
                newCountry('Argentina', printCountries);
            }, 3000);
        }, 2000);
    }, 3000);
}

startCallbackHell();