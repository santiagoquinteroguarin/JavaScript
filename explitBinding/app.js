// Explit Binding

function person(product1, product2) {
    console.log(`Mi nombre es: ${this.name} y me gusta el ${product1} y las ${product2}`);
}

const info = {
    name: 'Santiago',
}

const fruits = ['Mango','Fresas'];

// call - recibe de a un argumento
person.call(info, fruits[0], fruits[1]);
// apply - recibe un array
person.apply(info, fruits);
// bind - crea un nueva funci√≥n y recibe de a un argumento
const newFunction = person.bind(info, fruits[0], fruits[1]);
newFunction();