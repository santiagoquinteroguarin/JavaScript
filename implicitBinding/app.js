const person = {
    name: 'Santiago',
    age: 23,
    info() {
        console.log(`Mi nombre es ${this.name} y mi edad es ${this.age}`);
    },
    pet: {
        name: 'Luna',
        age: 5,
        info() {
            console.log(`Mi mascota se llama ${this.name} y tiene una edad de ${this.age} a√±os`);
        }
    }
}

person.info();
person.pet.info();