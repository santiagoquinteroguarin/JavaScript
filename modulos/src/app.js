import { Client } from './client.js';
import welcome, { Fruit } from './fruit.js';

welcome();
const client = new Client('Santigo');
console.log(client)
const fruit = new Fruit('Apple','🍎');
console.log(fruit)
