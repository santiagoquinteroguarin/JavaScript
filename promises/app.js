const iAmProgrammer = new Promise((resolve, reject) => {
    const validation = true;
    if(validation) {
        resolve('Soy programad@r ðŸ˜‰');
    } else {
        reject('No soy un programad@r ðŸ˜£');
    }
})

iAmProgrammer
    .then(resolve => console.log(resolve))
    .catch(err => console.log(err));