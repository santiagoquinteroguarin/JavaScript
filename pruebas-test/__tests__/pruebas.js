// ? empezando con jest
// un grupo de pruebas, sino quieres crear varios archivos para pruebas, creas un grupo 
describe('grupo de pruebas', () => {
    // first test - recibe un callback
    test('Hola Mundo!' ,() => {});
    // second test - recibe un callback
    it('Esta es otra forma de hacer un test', () => {});
});


//? pruebas para strings
const password = "123456";
describe('grupo de pruebas para strings, validar la longitud de password', () => {
    test('validar longitud', () => {
        expect(password).toHaveLength(6);
    });
    
    test('password no vacio', () => {
        expect(password).not.toHaveLength(0);
    });
});

// ? pruebas para arrays
const carrito = ['product1','product2','product3'];
describe('testing al carrito de compras', () => {
    test('probar que el arrays tenga 3 elementos', () => {
        expect(carrito).toHaveLength(3);
    })
    
    test('varificar que el carrito no este vacio', () => {
        expect(carrito).not.toHaveLength(0);
    })
});

// ? pruebas para objetos
const cliente = {
    name: 'Santiago',
    dinero: 500,
}

describe('Testing al cliente', () => {
    test('El cliente es premium, si un valor es mayor', () => {
        expect(cliente.dinero).toBeGreaterThan(400);
    })
    
    test('es Santiago', () => {
        expect(cliente.name).toBe('Santiago');
    });
    
    test('No, es otro cliente', () => {
        expect(cliente.name).not.toBe('Camilo')
    });
    
    test('No tiene 400', () => {
        expect(cliente.dinero).not.toBe(400)
    });
    
});

// ? pruebas para funciones
function suma(a, b) {
    return a + b;
}

function resta(a, b) {
    return a - b;
}

describe('Testing a funciones', () => {
    test('suma de 20 y 30', () => {
        expect(suma(20,30)).toBe(50);
    });
    
    test('resta de 50 y 20', () => {
        expect(resta(50,20)).toBe(30)
    });

    test('que la suma de 20 y 20, no sea 10', () => {
        expect(suma(20,20)).not.toBe(10);
    });

    test('que la resta de 5 y 20, no sea 30', () => {
        expect(resta(5,20)).not.toBe(35)
    });
    
});

// ? Snapshots - almacena una data como una bd y empieza a realizar pruebas con esta data
// para actualizar un snapshot npm test -- -u
const persona = {
    name: 'Santiago',
    balance: 400,
    type: 'premium',
}

describe('testing a persona', () => {
    test('Es Santiago', () => {
        expect(persona).toMatchSnapshot();
    });
    
});

// ? importar funciones, clases etc.. para hacer pruebas - para utilizar esto debemos instalar babel
// ? crear archivo .babelrc en la raiz y instalar la dependencia de desarrollo  npm i --save-dev @babel/preset-env
import { multi } from '../js/app.js';

describe('testing a la funcion multiplicar', () => {
    test('validacion de multiplicaciÃ³n 5 * 5 = 25', () => {
        expect(multi(5,5)).toBe(25);
    })
});






