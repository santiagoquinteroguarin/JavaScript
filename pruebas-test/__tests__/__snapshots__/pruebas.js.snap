// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing a persona Es Santiago 1`] = `
Object {
  "balance": 400,
  "name": "Santiago",
  "type": "premium",
}
`;
